### PR TITLE
Added Azure.MariaDB.FirewallRuleCount

### LINKED ISSUE #1855: Azure Database for MariaDB servers should restrict an excessive number of firewall rules
# Rule request

## Suggested rule change

Azure Database for MariaDB servers should restrict an excessive number of firewall rules.

Security pillar for this one.

## Applies to the following

The rule applies to the following:

- Resource type: **[Microsoft.DBforMariaDB/servers]**

## Additional context

[Azure Database for MariaDB server firewall rules](https://learn.microsoft.com/azure/mariadb/concepts-firewall-rules)
[Template reference](https://learn.microsoft.com/azure/templates/microsoft.dbformariadb/servers)

### PR DESCRIPTION
**Pull request by BenjaminEngeset**
## PR Summary

Fixes #1855

Added Azure.MariaDB.FirewallRuleCount rule.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [ ] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

**Comment by BenjaminEngeset** [comment]
Hey @BernieWhite 

- Could you tell me what you're thinking about using this logic? The rule is not bulletproof as it only runs against child resources declared within the parent resource, but from my experience most customers declare child resources within the parent resource anyways.
- I'm waiting for the merge on `Azure.MariaDB.AllowAzureAccess` that will help with good test cases.
- Should we add ARM and Bicep code examples for this one? Where we declare some firewall rules on a MariaDB server?

File: docs/en/rules/Azure.MariaDB.FirewallRuleCount.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+## DESCRIPTION
+
+Typically the number of firewall rules required is minimal, with management connectivity from on-premises and cloud application connectivity.
+
+Server-level firewall rules apply to all databases on the Azure Database for MariaDB server.
+
+## RECOMMENDATION
+
+Review Azure for MariaDB server firewall rules.
```
Comment (BernieWhite) [inline docs/en/rules/Azure.MariaDB.FirewallRuleCount.md:23]: I think we need a clearer message here about why the rule failed and what a customer should do.

**Review by BernieWhite — COMMENTED** [review]
@BenjaminEngeset I have a few comments below but it looks good so far.

For counting the number of firewall rules for the server, this logic looks good.
PSRule will automatically nest sub-resource into the parent resource if they exist in the same deployment/ template when expanding a template. So declaring sub-resources outside of the parent resource is not a concern.

You could write a template with only one firewall rule in it but since we can't make an assumption about how many rules already exist; it's an edge case we don't need to solve in this rule.

**Comment by BenjaminEngeset** [comment]
> @BenjaminEngeset I have a few comments below but it looks good so far.
> 
> > Could you tell me what you're thinking about using this logic? The rule is not bulletproof as it only runs against child resources declared within the parent resource, but from my experience most customers declare child resources within the parent resource anyways.
> 
> For counting the number of firewall rules for the server, this logic looks good. PSRule will automatically nest sub-resource into the parent resource if they exist in the same deployment/ template when expanding a template. So declaring sub-resources outside of the parent resource is not a concern.
> 
> You could write a template with only one firewall rule in it but since we can't make an assumption about how many rules already exist; it's an edge case we don't need to solve in this rule.
> 
> > Should we add ARM and Bicep code examples for this one? Where we declare some firewall rules on a MariaDB server?
> 
> Because this rule is about the number of firewall rules and reducing them, I don't think we need to an example code.

Thanks @BernieWhite. I had no idea that PSRule was that powerful in terms of automatically nesting sub-resources that are declared outside the parent resource, always taught they would been treated as standalones then.

**Comment by BenjaminEngeset** [comment]
@BernieWhite 

Check the latest updates and let me know you're thinking.

**Review by BernieWhite — APPROVED** [review]
Thanks @BenjaminEngeset. All good to merge.

**Pull request merged**